### PR TITLE
Lexically schedule to reduce feature-map memory requirements

### PR DESCRIPTION
**Pull request by ljfitz**
This pass re-orders the XTen operations to reduce the total size of the live tensors when executing sequentially.

- [x] add tests

### CODE REVIEW

File: lib/Transform/XTenMinimizeLiveTensors.cpp [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        continue;
+
+      op->moveBefore(fwd.op);
+      visitFwd.ordered = true;
+      moveToOrder(visitFwd);
+    }
+  }
+
+  void runOnOperation() override {
+    // TODO: no need to use a module pass here - a func pass would be ideal.
```
Comment (ghost) [inline lib/Transform/XTenMinimizeLiveTensors.cpp]:
```suggestion
```

remove comment

File: lib/Transform/XTenMinimizeLiveTensors.cpp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+  SmallVector<Value> const operands;
+  /// The feature map results this node.
+  SmallVector<Value> const results;
+  /// The value that will share memory with the result during execution, if any.
+  Optional<Value> const sharesResultMemory = {};
+  /// Cumulative sizes of the FM tensors.
+  OpSizes sizes = {};
+  /// The consumers of any results. Note: this is filled progressively while
+  /// collecting the operations.
+  SmallVector<Operation *> consumers;
```
Comment (ljfitz) [inline lib/Transform/XTenMinimizeLiveTensors.cpp:68]: Move up, as it is built in initial collection walk.

File: lib/Transform/XTenMinimizeLiveTensors.cpp [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+  if (isa<xilinx::xten::Conv2dTensorAddReLUOp>(op))
+    return {op->getOperands().back()};
+  if (isa<xilinx::xten::Conv2dTensorAddLReLUOp>(op))
+    return {op->getOperands().back()};
+  if (isa<xilinx::xten::Conv2dTensorAddGlobalAveragePoolOp>(op))
+    return {op->getOperands().back()};
+  if (isa<xilinx::xten::Conv2dTensorAddReLUGlobalAveragePoolOp>(op))
+    return {op->getOperands().back()};
+  if (isa<xilinx::xten::Conv2dTensorAddLReLUGlobalAveragePoolOp>(op))
+    return {op->getOperands().back()};
```
Comment (ljfitz) [inline lib/Transform/XTenMinimizeLiveTensors.cpp]: Make a function `isConvAddChained()` and use here and in `getFmOperands()`.

File: lib/Transform/XTenMinimizeLiveTensors.cpp [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    SmallVector<Value> const retVal = returnStmt->getOperands();
+    OpInfo fwdInfo = {.op = returnStmt, .operands = retVal, .results = {}};
+    auto [opFwdIt, succeeded] =
+        opToInfo.emplace(returnStmt, std::move(fwdInfo));
+    OpInfo const &retFwd = opFwdIt->second;
+
+    collectOperandInfo(retFwd);
+
+    setFwdSizes();
```
Comment (ljfitz) [inline lib/Transform/XTenMinimizeLiveTensors.cpp]: Move into `collectOperandInfo` - having it separate is not any clearer.

File: lib/Transform/XTenMinimizeLiveTensors.cpp [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    for (BranchRunning const &branch : branches)
+      maxRunning = std::max(maxRunning, branch.maxRunning);
+    brInfo.maxRunning = maxRunning;
+    brInfo.lastResults = opInfo->sizes.results;
+    brInfo.lastOp = opInfo->op;
+    completed.insert({opInfo->op, brInfo});
+
+    // Continue to all consumers.
+    for (Operation *consumer : opInfo->consumers) {
+      BranchRunning nextRunning{.maxRunning = maxRunning,
```
Comment (ljfitz) [inline lib/Transform/XTenMinimizeLiveTensors.cpp:268]: This is propagating too large a `maxRunning` into the innermost branching choices, were e.g. the `running` of the current `op` may dominate all consumer `running` and break the sorting within the inner branch-join graphs.

**Review by ghost — APPROVED** [review]

File: lib/Transform/XTenMinimizeLiveTensors.cpp (hunk 5 above)
Comment (ljfitz) [reply lib/Transform/XTenMinimizeLiveTensors.cpp:268]: @ehsan-toosi - I'll address this in a follow-up PR.

**Review by ehsan-toosi — APPROVED** [review]

**Pull request merged**